compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs

```diff
@@ -283,86 +298,153 @@ generateTemplateInstance env typeCon typeParams externPkgId =
-                                                                           NoSourceText $
-                                                                       mkFastString
-                                                                           ((T.unpack $
-                                                                             LF.unPackageId
-                                                                                 externPkgId) <>
-                                                                            ":" <>
-                                                                            moduleNameStr <>
-                                                                            ":" <>
-                                                                            (T.unpack $
-                                                                             T.intercalate
-                                                                                 "." $
-                                                                             LF.unTypeConName
-                                                                                 templName) <>
-                                                                            ":" <>
-                                                                            T.unpack
-                                                                                funName))))
-                                                      ]
-                                                , grhssLocalBinds =
-                                                      noLoc emptyLocalBinds
-                                                }
-                                      }
-                                ]
-                      , mg_origin = Generated
+
+-- | Generate a single choice instance for a given template/choice
+generateChoiceInstance ::
+       Env
+    -> LF.PackageId
+    -> LF.Template
+    -> LF.TemplateChoice
+    -> HsDecl GhcPs
+generateChoiceInstance env externPkgId template choice =
+    InstD noExt $
+    ClsInstD
+        noExt
+        ClsInstDecl
+            { cid_ext = noExt
+            , cid_poly_ty =
+                  HsIB
+                      { hsib_ext = noExt
+                      , hsib_body = body
                       }
-            , fun_co_fn = WpHole
-            , fun_tick = []
+            , cid_binds = listToBag $ map classMethodStub methodMapping
+            , cid_sigs = []
+            , cid_tyfam_insts = []
+            , cid_datafam_insts = []
+            , cid_overlap_mode = Nothing
             }
+  where
+    app :: LHsType GhcPs -> LHsType GhcPs -> LHsType GhcPs
+    app t1 t2 = noLoc $ HsAppTy noExt t1 t2
+
+    body :: LHsType GhcPs =
+      choiceClass `app` arg1 `app` arg2 `app` arg3
+
+    choiceClass :: LHsType GhcPs =
+        noLoc $
+        HsTyVar noExt NotPromoted $
+        noLoc $
+        mkRdrQual (mkModuleName "DA.Internal.Template") $
```

Comment:
You'll also need to qualify this as `Sdk.DA.Internal.Template`.

Reply:
same as above